ooniprobe 3.1.0

hellais opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
+  license "BSD-3-Clause"
+  revision 1
 
-  bottle do
-    cellar :any
-    sha256 "9a5d8c8b6bda3609642113631ba7c39b2cbf4fc27b09bd4b2fccc832befdd3e5" => :catalina
-    sha256 "e8e120b4342f22d48efbcfa45cde2faa28c9edd045121373f3b2ba8349e1d6fc" => :mojave
-    sha256 "3e13549c0175e9f3167f24526ed0c45bd7096b84c0360042654be9b4dff980f7" => :high_sierra
-  end
```
Please don't remove the bottle block

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
-    url "https://files.pythonhosted.org/packages/44/af/cea1e18bc0d3be0e0824762d3236f0e61088eeed75287e7b854d65ec9916/zope.interface-4.3.3.tar.gz"
-    sha256 "8780ef68ca8c3fe1abb30c058a59015129d6e04a6b02c2e56b9c7de6078dfa88"
-  end
+  depends_on "go@1.14" => :build
```
Does it not compile with the latest Go version?

hellais replied:
It does build also with the latest version of golang, but we do extensive testing with this specific version.

@bassosimone can maybe add more details about this aspect.

bassosimone replied:
It does compile with Go 1.15. Yet, with Go 1.15, one of our dependencies (the psiphon censorship circumvention system; https://github.com/Psiphon-Labs/psiphon-tunnel-core) does not support using QUIC for censorship evasion.

Psiphon will migrate to Go 1.15 when Go 1.14 is going EOL. Currently Psiphon builds its own apps and generally always uses Go 1.14 for its circumvention system.

For this reason, we're using Go 1.14 such that our testing of Psiphon is closer to how Psiphon is usually used (and specifically such that it can use QUIC as an evsasion mechanism).

bassosimone replied:
Would it be okay for us adding a comment explaining why we're using Go 1.14?

SMillerDev replied:
If it works with go 1.15 it should use go 1.15, such are the implications of a evergreen package-manager like Homebrew

hellais replied:
@SMillerDev using go 1.15 leads to our upstream dependencies breaking and some missed functionality. 

As soon our upstream fixes this, we will be sure to further revise the formula to run with golang 1.14.

Is it OK in this case to keep this as-is?

SMillerDev replied:
Since it's essentially a new formula I'd like for the formula to use the latest go version before merging. This way we avoid adding the formula with some bit rot build in.

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
-    ENV.append "CPPFLAGS", "-I#{MacOS.sdk_path}/usr/include/ffi" if MacOS.sdk_path_if_needed
-
-    virtualenv_install_with_resources
+    ENV["GOPATH"] = HOMEBREW_CACHE/"go_cache"
```
This shouldn't need to be set anymore

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
+    cd "src/github.com/ooni/probe-cli" do
+      system "./build.sh macos"
+      bin.install "CLI/darwin/amd64/ooniprobe"
+      prefix.install_metafiles
```
```suggestion
```

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
+    ooni_home = Pathname.new "#{var}/ooniprobe"
+    ooni_home.mkpath
+
+    (prefix/"share/daily-config.json").write <<~EOS
```
This should be distributed as part of the package.

hellais replied:
We would like to ship this specific configuration file as part of homebrew, because we don't want to bypass the informed consent procedure as part of the default install.

We are going to think of an alternative way to achieve this without needing to include a config file using this pattern.

SMillerDev replied:
Why would homebrew users not need to give consent when other parties should?

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
-
-    ln_s pkgshare/"decks", pkgshare/"decks-available"
-    ln_s pkgshare/"decks/web.yaml", pkgshare/"current.deck"
+    ooni_home = Pathname.new "#{var}/ooniprobe"
+    ooni_home.mkpath
```
This should be done in the install

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
+
+          <key>EnvironmentVariables</key>
+          <dict>
+            <key>PATH</key>
+            <string>#{HOMEBREW_PREFIX}/bin:/usr/bin:/bin:/usr/sbin:/sbin</string>
```
Why should it alter the PATH?

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -3,287 +3,163 @@ class Ooniprobe < Formula
-    system bin/"ooniprobe", "-ng", "blocking/tcp_connect", "-f", testpath/"ooni/hosts.txt"
+    mkdir_p "#{testpath}/ooni_home"
+    ENV["OONI_HOME"] = "#{testpath}/ooni_home"
+    system bin/"ooniprobe", "--config", testpath/"config.json", "run", "websites"
```
Can we check some output here?

hellais commented:
Thanks for your review @SMillerDev we are going to work on another iteration of this and ping you when it's ready for review.

hellais commented:
@SMillerDev we have addressed all your feedback (except for the golang 1.14 issue which we would like your advice on).

Let us know if any further changes are necessary to get this merged.

Thanks a bunch!

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -15,275 +15,108 @@ class Ooniprobe < Formula
+          <key>ProgramArguments</key>
+          <array>
+              <string>#{opt_bin}/ooniprobe</string>
+              <string>--log-handler=syslog</string>
```
Why specifically syslog?

bassosimone replied:
Because syslog directs the logs to Apple's logger. So a user can open /Applications/Utility/Console.app and filter for `ooniprobe` to observe logs generated by us. But we don't write files on disk and we comply as much as possible to what seem Apple expectations from developers (AFAICT, but correct me if wrong). The specific implementation is using libc's `syslog(3)`, which seem to do the right thing for our user case (i.e., logs safely land in Console.app).

SMillerDev replied:
Is this the way a normal user would run `ooniprobe`? I personally wouldn't expect non-system programs to show up in the Console app.

hellais replied:
So that it ends up inside of the Console app in macOS and doesn't leave files on disk, potentially occupying a lot of on-disk storage.

bassosimone replied:
A normal user runs `ooniprobe run [...]` from the console and gets messages on the console. This part of the diff is creating a `.plist` such that a user could configure `brew services start ooniprobe` as a launch agent that is periodically run in the background in the context of the user session. The choice to log on the systlog, therefore, only pertains to the use case where we're running in the background.

SMillerDev replied:
I understand that this is a service, but that should be running as the default state that upstream defined regardless of your views on disk usage by homebrew formulae. If upstream says "make a systemd unit with `/usr/bin/x -a -b`" then the plist file should also be made that way.

bassosimone replied:
We are the upstream (both @hellais and me are core developers of @ooni). This is the recommended way in which we want people to run ooniprobe when running in the background :-).

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -15,275 +15,108 @@ class Ooniprobe < Formula
-      Decks are installed to #{opt_pkgshare}/decks.
+      By enabling the homebrew service you will not be shown the informed consent.
+
+      WARNING:
+
+      • OONI Probe will likely test objectionable sites and services
+      • Anyone monitoring your internet activity (such as your government
+        or Internet provider) may be able to tell that you are using OONI Probe
+      • The network data you collect will be published automatically
+
+      To learn more about the risks see:
+          https://ooni.org/about/risks
```
This should be in upstream documentation as it is not brew specific

hellais replied:
It's also in our upstream documentation, but we would like to display some message to users informing them of the potential risks since they are bypassing the traditional informed consent procedure that they go through when installing the apps through canonical channels.

Is there some way to provide some information text to end users and/or prompt in homebrew?

SMillerDev replied:
Yes, make it part of the app 😄

hellais replied:
We have it as part of the app, however this is information text for the background service which bypasses that.

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -15,275 +15,108 @@ class Ooniprobe < Formula
   def caveats
     <<~EOS
-      Decks are installed to #{opt_pkgshare}/decks.
+      By enabling the homebrew service you will not be shown the informed consent.
```
Why not actually?

hellais replied:
Because it's running in the background and there is no way to show a prompt to users.

SMillerDev replied:
If it doesn't run without consent then that's fine too. I don't think skipping consent is a good price to pay for an easier first run.

I get the impression ooniprobe should always be ran as a service, so are you skipping the consent for all other install methods? Because then it might be better to remove it altogether.

SMillerDev commented on Formula/ooniprobe.rb:
```diff
@@ -15,275 +15,108 @@ class Ooniprobe < Formula
-
-    ln_s pkgshare/"decks", pkgshare/"decks-available"
-    ln_s pkgshare/"decks/web.yaml", pkgshare/"current.deck"
+    ENV["OONI_HOME"] = "#{HOMEBREW_PREFIX}/var/ooniprobe"
```
If it doesn't work without this variable the binary should be wrapped using `write_env_script` to always set it.

hellais replied:
This is only needed to write the configuration file for the background launchd system unit. It's not needed by the normal CLI tool which will have a different OONI_HOME (the default set to `~/.ooniprobe`).

hellais commented:
We discussed this with @bassosimone and we are going to take a different approach. Instead of relying on homebrew for installing the launchd service, we are going to implement in ooniprobe the capability to configure and launch the service directly.

This should significantly simplify the homebrew package, as well as avoid bypassing the standard informed consent procedure.

Thanks for your thorough review, we will update you when it's ready for review again.

hellais commented:
We further discussed this and came up with a simpler solution to this problem. We are going to work on this in the current sprint and aim to update this PR within a few weeks.

SMillerDev commented:
Please create a new one with the new solution then
